perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/VosManagerBlImpl.java

```diff
@@ -902,14 +929,65 @@ public boolean usesEmbeddedGroupRegistrations(PerunSession sess, Vo vo) {
+
+		List<Member> memberVoMembers = perunBl.getMembersManagerBl().getMembers(sess, memberVo);
+		AttributeDefinition memberExpirationAttrDef;
+
+		try {
+			memberExpirationAttrDef = perunBl.getAttributesManagerBl().getAttributeDefinition(sess, AttributesManager.NS_MEMBER_ATTR_DEF + ":membershipExpiration");
+		} catch (AttributeNotExistsException e) {
+			throw new InternalErrorException(e);
+		}
+
+		for (Member member : memberVoMembers) {
+			try {
+
+				try {
+					// if user is member in both vos, update memberOrganizations list attribute, validate and reset expiration
+					Member existingMember = perunBl.getMembersManagerBl().getMemberByUserId(sess, vo, member.getUserId());
+					perunBl.getAttributesManagerBl().removeAttribute(sess, existingMember, memberExpirationAttrDef);
+					if (existingMember.getStatus().equals(Status.INVALID) || existingMember.getStatus().equals(Status.DISABLED)) {
+						perunBl.getMembersManagerBl().validateMemberAsync(sess, member);
+					}
+					Attribute attribute = perunBl.getAttributesManagerBl().getAttribute(sess, existingMember, A_MEMBER_DEF_MEMBER_ORGANIZATIONS);
+					ArrayList<String> currentValue = attribute.valueAsList();
+					currentValue.add(memberVo.getShortName());
+					attribute.setValue(currentValue);
+					perunBl.getAttributesManagerBl().setAttribute(sess, existingMember, attribute);
+
+				} catch (MemberNotExistsException e) {
+					// if user is member only in member vo, create member in parent vo and set up memberOrganizations list attribute
+					Member newMember = perunBl.getMembersManagerBl().createMember(sess, vo, perunBl.getUsersManagerBl().getUserByMember(sess, member));
+					perunBl.getMembersManagerBl().validateMemberAsync(sess, member);
+					Attribute attribute = new Attribute(perunBl.getAttributesManagerBl().getAttributeDefinition(sess, A_MEMBER_DEF_MEMBER_ORGANIZATIONS));
+					ArrayList<String> newValue = new ArrayList<>(List.of(memberVo.getShortName()));
+					attribute.setValue(newValue);
+					perunBl.getAttributesManagerBl().setAttribute(sess, newMember, attribute);
+				}
+
+			} catch (WrongAttributeAssignmentException | WrongReferenceAttributeValueException | AttributeNotExistsException | WrongAttributeValueException | AlreadyMemberException | ExtendMembershipException e) {
+				throw new InternalErrorException(e);
+			}
+		}
 	}
 
 	@Override
 	public void removeMemberVo(PerunSession sess, Vo vo, Vo memberVo) throws RelationNotExistsException {
-		// todo - add necessary logic for removing member vo
 		vosManagerImpl.removeMemberVo(sess, vo, memberVo);
+
+		List<Member> parentVoMembers = perunBl.getMembersManagerBl().getMembers(sess, vo);
+
+		for (Member member : parentVoMembers) {
+			try {
+				Attribute attribute = perunBl.getAttributesManagerBl().getAttribute(sess, member, A_MEMBER_DEF_MEMBER_ORGANIZATIONS);
+				ArrayList<String> currentValue = attribute.valueAsList();
+				attribute.setValue(currentValue.remove(vo.getShortName()));
```

Comment:
This all seems confusing to me. If we're removing memberVo from the hierarchy, don't we want to remove memberVo from the memberOrganizations attribute? If that's the case, I believe we only do this with members that are from memberVo, not all parentVo members.